### PR TITLE
Replace the context cancellation on signal with signal.NotifyContext

### PR DESCRIPTION
**Pull request by ryysud**
Signed-off-by: Ryuma Yoshida <ryumyosh@zlab.co.jp>

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [ ] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**

`spire-server run` and `spire-agent run` commands.

**Description of change**

I replaced the context cancellation on signal with signal.NotifyContext.

**Which issue this PR fixes**

A part of https://github.com/spiffe/spire/issues/2222.

### CODE REVIEW

File: pkg/common/telemetry/inmem.go [hunk 1]
```diff
@@ -78,20 +77,20 @@ func (i *inmemRunner) run(ctx context.Context) error {
 	return nil
 }
 
-func (i *inmemRunner) startConfigWarning(ctx context.Context, wg *sync.WaitGroup) {
+func (i *inmemRunner) startConfigWarning(parent context.Context, wg *sync.WaitGroup) {
 	wg.Add(1)
-	sigChannel := make(chan os.Signal, 1)
-	signal.Notify(sigChannel, metrics.DefaultSignal)
+	ctx, cancel := signal.NotifyContext(parent, metrics.DefaultSignal)
```
Comment (azdagron) [inline pkg/common/telemetry/inmem.go:82]: This change is also incorrect since it will lead to a for-select spin loop with the cancelled context. The current code is correct.

Can we revert this change?

**Review by azdagron — COMMENTED** [review]
Thank you for opening this @ryysud! I think there is some confusion though. We don't want to replace all the uses of `signal.Notify` with one that cancels a context. We only want to replace those that are currently using signal.Notify + a goroutine to cancel a context already (e.g. callers of util.SignalListener).

File: pkg/common/log/reopen_posix.go [hunk 2]
```diff
@@ -16,24 +15,25 @@ const (
 
 // ReopenOnSignal returns a function compatible with RunTasks.
 func ReopenOnSignal(logger *Logger, reopener Reopener) func(context.Context) error {
-	return func(ctx context.Context) error {
-		signalCh := make(chan os.Signal, 1)
-		signal.Notify(signalCh, reopenSignal)
-		return reopenOnSignal(ctx, logger, reopener, signalCh)
+	return func(parent context.Context) error {
+		ctx, cancel := signal.NotifyContext(parent, reopenSignal)
+		return reopenOnSignal(parent, ctx, cancel, logger, reopener)
```
[the remark this answers is not in the document]
Comment (ryysud) [reply pkg/common/log/reopen_posix.go]: https://github.com/spiffe/spire/pull/2901/commits/5df363bb1503ceccdb82eeebf44210fa5284c8ac

File: pkg/common/telemetry/inmem.go (hunk 1 above)
Comment (ryysud) [reply pkg/common/telemetry/inmem.go]: https://github.com/spiffe/spire/pull/2901/commits/5df363bb1503ceccdb82eeebf44210fa5284c8ac

File: pkg/common/util/signals.go [hunk 3]
```diff
@@ -2,20 +2,17 @@ package util
 
 import (
 	"context"
-	"os"
 	"os/signal"
 	"syscall"
 )
 
-func SignalListener(ctx context.Context, cancel func()) {
+func SignalListener(parent context.Context, cancel func()) {
```
[the remark this answers is not in the document]
Comment (ryysud) [reply pkg/common/util/signals.go:9]: Thank you for your review, @azdagron!
I understand the background of the issue, I will fix it next week.

File: pkg/common/util/signals.go [hunk 4]
```diff
@@ -2,20 +2,17 @@ package util
...
 	go func() {
-		signalCh := make(chan os.Signal, 1)
-		signal.Notify(signalCh, syscall.SIGINT, syscall.SIGTERM)
+		ctx, cancel := signal.NotifyContext(parent, syscall.SIGINT, syscall.SIGTERM)
 
 		select {
+		case <-parent.Done():
```
Comment (azdagron) [inline pkg/common/util/signals.go:14]: This select case is unnecessary since if `parent` is cancelled, `ctx` will be cancelled too.

File: pkg/common/util/signals.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/common/util/signals.go:9]: Wonderful, thanks!

File: pkg/common/util/signals.go [hunk 5]
```diff
@@ -2,20 +2,17 @@ package util
...
 		case <-ctx.Done():
-			return
-		case <-signalCh:
 			cancel()
```
Comment (azdagron) [inline pkg/common/util/signals.go:16]: Line 11 now declares a `cancel` variable that is going to mask the one provided in the arguments, which will prevent this function from cancelling the parent context when a signal is received. Both the cancel in the argument and the cancel for the child context need to be called in order to cancel the parent and clean up resources.

**Comment by azdagron** [comment]
Hi @ryysud ! Just checking in to see if you need anything.

File: pkg/common/util/signals.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryysud) [reply pkg/common/util/signals.go]: Fix: https://github.com/spiffe/spire/pull/2901/commits/9909123e32936a07dc8284761bf42f01e2e4d345

File: pkg/common/util/signals.go (hunk 4 above)
Comment (ryysud) [reply pkg/common/util/signals.go]: I removed `utils.SignalListener`.

File: pkg/common/util/signals.go (hunk 5 above)
Comment (ryysud) [reply pkg/common/util/signals.go]: I removed `utils.SignalListener`.

**Review by azdagron — APPROVED** [review]
Thanks so much, @ryysud !
